src/main/java/net/javadiscord/javabot/util/IndentationHelper.java

```diff
@@ -0,0 +1,205 @@
+		/**
+		 * Two spaces should be used for indentation.
+		 */
+		TWO_SPACES("  "),
+		/**
+		 * Doesn't define an indentation type but
+		 * Is used as a substitute to null to indicate that a String given to any method in {@link IndentationHelper} should not be changed.
+		 */
+		NULL("");
+
+		/**
+		 * Holds the characters used for indentation of the type.
+		 */
+		private final String pattern;
+
+		/**
+		 * Constructs the indentation type.
+		 *
+		 * @param pattern The pattern to be used as indentation
+		 */
+		IndentationType(String pattern) {
+			this.pattern = pattern;
+		}
+
+		/**
+		 * Get the pattern for a given Indentation type.
+		 *
+		 * @return the pattern to be used for indenting.
+		 */
+		public String getPattern() {
+			return pattern;
+		}
+
+		/**
+		 * Returns the number of characters a pattern is using.
+		 *
+		 * @return the number of characters the pattern of this type consists of.
+		 */
+		private int getNumberOfChars() {
+			return pattern.length();
+		}
+	}
+
+	/**
+	 * Private enum to denote the current State of the Indentation Process.
+	 */
+	private enum IndentationState {
+		/**
+		 * Denotes that the Process is currently in a codeblock and should indent the code accordingly.
+		 */
+		CODE,
+		/**
+		 * Denotes that the Process is currently in a String literal.
+		 */
+		STRING,
+		/**
+		 * Denotes that the Indentation Process is currently in a character literal.
+		 */
+		CHARACTER,
+		/**
+		 * Denotes that the Process is currently in a single line comment.
+		 */
+		SINGLE_LINE_COMMENT,
+		/**
+		 * Denotes that the process is inside a multi line comment or a javadoc.
+		 */
+		MULTI_LINE_COMMENT
+	}
+
+	private final String text;
+	private final IndentationType type;
+	private boolean startOfLine = true;
+	private int numberOfBrackets = 0;
+	private StringBuilder builder;
+
+	private IndentationHelper(String text, IndentationType type) {
+		this.text = text;
+		this.type = type;
+		this.builder = new StringBuilder((int) (text.length() * 1.25f));
+	}
+
+	/**
+	 * Aims to indent the given String using the pattern provided. Will return the String unchanged if {@link IndentationHelper.IndentationType#NULL} is passed as the IndentationType parameter.
+	 *
+	 * @param text The text that should be indented.
+	 * @param type The type of indentation to be used.
+	 * @return The indented String with the format specified.
+	 */
+	public static String formatIndentation(String text, IndentationType type) {
+		return new IndentationHelper(text, type).formatIndentation();
+	}
+
+	private String formatIndentation() {
+		if (type == IndentationType.NULL) {
+			return text;
+		}
+		IndentationState currentState = IndentationState.CODE;
+		for (int i = 0; i < text.length(); i++) {
+			char current = text.charAt(i);
+			if (startOfLine && current == ' ') {
+				continue;
+			}
+			builder.append(current);
+			if (current == '\n') {
+				builder.append(type.getPattern().repeat(Math.max(numberOfBrackets, 0)));
```

Comment:
If there are whitespaces at the start of the line already, are they preserved? Wouldn't that introduce redundant whitespaces? (If yes, please check the code from earlier for a possible mitigation).

Reply:
Whitespaces at the start of a line should not be preserved. If they are please let me know because that is definitely a bug.